parachain/src/wasm_executor/validation_host.rs

```diff
@@ -39,6 +38,37 @@ const NUM_HOSTS: usize = 8;
 /// Execution timeout in seconds;
 pub const EXECUTION_TIMEOUT_SEC: u64 =  5;
 
+#[derive(Default)]
+struct WorkerExternalitiesInner {
+	egress_data: Vec<u8>,
+	up_data: Vec<u8>,
+}
+
+#[derive(Default, Clone)]
+struct WorkerExternalities {
+	inner: Arc<Mutex<WorkerExternalitiesInner>>,
```

Comment:
why does this need to be shared? I would have thought that every parachain execution can own its data completely.

Reply:
They are unique, that is correct. However, the design of the externalities extensions require that we pass an owned struct that implements `Send`. (This is related to how we store them internally).

The validation context needs to access the data written by the parachain afterwards, so we wrap the inner in a `Arc<Mutex<>>` to make it clonable and to make it implement `Send`.

Reply:
I really hope this is on the way out soon! It's polluting a bunch of stuff.